### PR TITLE
gh-106529: Make FOR_ITER a viable uop

### PR DESCRIPTION
**Pull request by gvanrossum**
Also clean up a few nits in the code generator, and add back an important Make dependency for ceval.o.

* Issue: gh-106529

### CODE REVIEW

**Comment by gvanrossum** [comment]
Hold on, need to remove merge conflicts from commit history.

File: Tools/cases_generator/flags.py [hunk 1]
```diff
@@ -176,7 +175,7 @@ def variable_used_unspecialized(node: parsing.Node, name: str) -> bool:
     tokens: list[lx.Token] = []
     skipping = False
     for i, token in enumerate(node.tokens):
-        if token.kind == "MACRO":
+        if token.kind == "CMACRO":
```
Comment (gvanrossum) [inline Tools/cases_generator/flags.py]: NOTE: This fix resulted in `_SPECIALIZE_UNPACK_SEQUENCE` becoming a viable uop. It was missing a `TIER_ONE_ONLY` marker; I've added it back. (The fix is needed to restore the feature that this function doesn't look inside `#if TIER_ONE`.)

**Comment by gvanrossum** [comment]
@brandtbucher Let me know how bad this interferes with the JIT branch. The stuff I put in `_FOR_ITER_TIER_TWO` probably shouldn't be compiled literally into the template. Maybe I should make part of the code at the `deoptimize` label into a macro? (But the setting of `frame->instr_ptr` must differ -- we can't use `target` since that's also used by error exits, since GH-112065.)

**Comment by gvanrossum** [comment]
Benchmark results:
https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20231115-3.13.0a1+-5c5d8bd-PYTHON_UOPS

**Comment by brandtbucher** [comment]
> @brandtbucher Let me know how bad this interferes with the JIT branch. The stuff I put in `_FOR_ITER_TIER_TWO` probably shouldn't be compiled literally into the template. Maybe I should make part of the code at the `deoptimize` label into a macro? (But the setting of `frame->instr_ptr` must differ -- we can't use `target` since that's also used by error exits, since [GH-112065](https://github.com/python/cpython/pull/112065).)

So, this doesn't actually look too bad. I think with a tiny bit of reworking, this could be merged into the JIT branch with no pain:

- Move the line that sets `frame->instr_ptr` from the `exit_trace` label into the `_EXIT_TRACE` instruction, using a macro (like `CURRENT_TARGET()` or something) to access the `target` member.
- Remove the stuff from `_FOR_ITER_TIER_TWO` that updates the stats, saves the stack pointer, and decrefs the executor, and just `goto exit_trace;` instead after updating `frame->instr_ptr`.

**Comment by gvanrossum** [comment]
Okay, I think I managed to do that. I can now merge into the justin branch with only a single trivial merge conflict (we both added something to the end of ceval_macros.h). Next up of course, how would I redefine CURRENT_TARGET() in templace.c?

**Comment by brandtbucher** [comment]
`#define CURRENT_TARGET() (target)`

**Comment by gvanrossum** [comment]
> `#define CURRENT_TARGET() (target)`

Yup, and I also needed to move the `exit_trace:` label one line down (to avoid setting `frame->instr_ptr` twice).

It seems to work, but unsure how to prove it (it compiles and passes tests, but it would too if the JIT was never invoked).

Anyway, we should probably wait until we've decided that making Tier 2 5% slower is a good idea.

**Comment by markshannon** [comment]
There seem to be a few unrelated changes in this PR.
Could you make another PR for the increase to the uop buffer size, extra debugging and change to `UNPACK_SEQUENCE`?

File: Python/bytecodes.c [hunk 2]
```diff
@@ -2535,6 +2536,28 @@ dummy_func(
...
+                if (_PyErr_Occurred(tstate)) {
+                    if (!_PyErr_ExceptionMatches(tstate, PyExc_StopIteration)) {
+                        GOTO_ERROR(error);
+                    }
+                    _PyErr_Clear(tstate);
+                }
+                /* iterator ended normally */
+                Py_DECREF(iter);
+                STACK_SHRINK(1);
+                /* HACK: Emulate DEOPT_IF to jump over END_FOR */
```
Comment (markshannon) [inline Python/bytecodes.c]: No hacks, please 🙂 

The code should look like this:
```
if (next == NULL) {
    if (_PyErr_Occurred(tstate)) {
        if (!_PyErr_ExceptionMatches(tstate, PyExc_StopIteration)) {
            GOTO_ERROR(error);
        }
        _PyErr_Clear(tstate);
    }
    /* iterator ended normally */
    Py_DECREF(iter);
    STACK_SHRINK(1);
    DEOPT_IF(true);
}
```

The trace generator can adjust the target, so it points after the `END_FOR`.

**Comment by gvanrossum** [comment]
> Could you make another PR for the increase to the uop buffer size, extra debugging and change to `UNPACK_SEQUENCE`?

Sure, see gh-112214. I'll just merge that once the tests pass, then merge it into this PR.

**Comment by gvanrossum** [comment]
Okay, here's the new version. @markshannon Want to review one more time?

File: Python/bytecodes.c [hunk 3]
```diff
@@ -4009,6 +4028,7 @@ dummy_func(
 
         op(_EXIT_TRACE, (--)) {
             TIER_TWO_ONLY
+            frame->instr_ptr = CURRENT_TARGET() + _PyCode_CODE(_PyFrame_GetCode(frame));
```
Comment (brandtbucher) [inline Python/bytecodes.c]: If you want to, you can revert this change (plus the ones in `ceval.c` and `ceval_macros.h`) since `_FOR_ITER_TIER_TWO` has been reworked.

File: Python/optimizer.c [hunk 4]
```diff
@@ -620,6 +621,12 @@ translate_bytecode_to_trace(
                                 }
                                 if (_PyUop_Replacements[uop]) {
                                     uop = _PyUop_Replacements[uop];
+                                    if (uop == _FOR_ITER_TIER_TWO) {
+                                        target += 1 + INLINE_CACHE_ENTRIES_FOR_ITER + oparg + 1;
+                                        _Py_CODEUNIT word = _PyCode_CODE(code)[target-1];
```
Comment (brandtbucher) [inline Python/optimizer.c]: `word` is unused on non-debug builds.

File: Tools/cases_generator/instructions.py [hunk 5]
```diff
@@ -115,7 +115,7 @@ def __init__(self, inst: parsing.InstDef):
     def is_viable_uop(self) -> bool:
         """Whether this instruction is viable as a uop."""
         dprint: typing.Callable[..., None] = lambda *args, **kwargs: None
-        if "FRAME" in self.name:
+        if self.name == "_FOR_ITER_TIER_TWO":
```
Comment (brandtbucher) [inline Tools/cases_generator/instructions.py]: Left-over debugging?

**Comment by gvanrossum** [comment]
I've applied Brandt's suggestions.

I've also split off some more cleanups and debug improvements into gh-112218 and gh-112220. Once those land I'll merge main once more and this should shrink a bit more (the cases_generator tweaks will be separated out).

**Comment by gvanrossum** [comment]
Benchmark says 4% slower (with Tier 2):
https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20231117-3.13.0a1+-14aea56-PYTHON_UOPS

**Comment by gvanrossum** [comment]
FWIW, I have a [branch](https://github.com/gvanrossum/cpython/tree/faster-uops) on top of this that makes the uop interpreter [3% faster](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20231119-3.13.0a1%2B-a2c4f00-PYTHON_UOPS/README.md). The approach is to let the code generator generate code that extracts `oparg` and `operand` from the instruction stream as needed (i.e., only for opcodes that use them) rather than the interpreter loop preemptively extracting them for every instruction. I'm not completely happy with it, and it probably would make life a little harder for the JIT template, which doesn't like directly referencing `next_uop` -- this should probably wait until we're ready to generate the JIT template code separately from the Tier 2 interpreter code.

**Comment by brandtbucher** [comment]
Honestly, as long as it’s done in macros, it shouldn’t be too bad. Macros are a great escape hatch in the short term before we start generating bespoke “JIT cases”.

**Comment by gvanrossum** [comment]
> Honestly, as long as it’s done in macros, it shouldn’t be too bad. Macros are a great escape hatch in the short term before we start generating bespoke “JIT cases”.

So it could generate something like
```
case _FOO_UOP: {
    oparg = CURRENT_OPARG();
    operand = CURRENT_OPERAND();
    ...
}
```
where the default definitions for those (in ceval_macros.h) would be
```
#define CURRENT_OPARG() next_uop[-1].oparg
#define CURRENT_OPERAND() next_uop[-1].operand
```

**Comment by brandtbucher** [comment]
Yup!
